apps/src/aichat/redux/slice.ts

```diff
@@ -55,6 +60,29 @@ const aichatSlice = createSlice({
     ) => {
       state.studentChatHistory = action.payload;
     },
+    setOwnChatHistory: (state, action: PayloadAction<ServerChatEvent[]>) => {
+      // Find the last index of an event that marks the start a new conversation with the model.
+      const events = action.payload;
+      let lastResetIndex = -1;
+      for (let i = events.length - 1; i >= 0; i--) {
+        if (
+          RESET_CONVERSATION_CUSTOMIZATION_UPDATES.includes(
+            (events[i] as ModelUpdate)?.updatedField
+          ) ||
+          (events[i] as UserActionEvent)?.descriptionKey === 'CLEAR_CHAT'
```

Comment:
Optional: you can use the type checker functions `isModelUpdate` and `isUserActionEvent` to narrow the types here to avoid casting; probably the same number of lines of code, but avoids the cast and optional chaining, e.g.

```
(isModelUpdate(event) && RESET_CONVERSATION_CUSTOMIZATION_UPDATES.includes(event.updatedField)) ||
          (isUserActionEvent(event) && event.descriptionKey === 'CLEAR_CHAT')
```